backend/server/src/nix/mod.rs

```diff
@@ -42,33 +63,44 @@ impl EvalService {
     }
 
     /// Given a drv, traverse all direct drv dependencies
-    fn traverse_drvs(&mut self, drv_path: &str) -> Result<()> {
-        if self.drv_map.contains_key(drv_path) {
+    async fn traverse_drvs(&mut self, drv_path: &str) -> Result<()> {
+        debug!("Entering traverse drvs");
+        if self.drv_map.contains_key(drv_path) || self.db_service.has_drv(drv_path).await? {
             debug!("Already evaluated {}, skipping....", drv_path);
             return Ok(());
         }
 
+        // This is used to collect drvs for insertion into the database
+        // We must know all of the drvs before refrencing relationships
+        // So we must complete the traversal, then attempt assertion of
+        // drvs (which are the keys in this case), then can add the references
+        let mut new_drvs: HashMap<String, Vec<String>> = HashMap::new();
+
         debug!("traversing {}", drv_path);
-        self.inner_traverse_drvs(drv_path)?;
+        self.inner_traverse_drvs(drv_path, &mut new_drvs)?;
+        self.db_service.insert_drv_graph(new_drvs).await?;
 
         Ok(())
     }
 
-    /// To avoid lifetime issues, we do a recursive descent instead of a loop
-    /// instead of appending to an iterator :(
-    fn inner_traverse_drvs(&mut self, drv_path: &str) -> Result<()> {
+    /// To avoid lifetime issues, we do a recursive descent
+    /// instead of appending to an iterator and a loop :(
+    fn inner_traverse_drvs(
+        &mut self,
+        drv_path: &str,
+        new_drvs: &mut HashMap<String, Vec<String>>,
+    ) -> Result<()> {
         let references = drv_references(drv_path)?;
-        // TODO: this drv hasn't been built before, it should eventually it put into a "new drv"
-        // queue
         debug!("new drv, traversing {}", &drv_path);
         self.drv_map
             .insert(drv_path.to_string(), references.clone());
+        new_drvs.insert(drv_path.to_string(), references.clone());
 
         for drv in references.into_iter() {
             if self.drv_map.contains_key(&drv) {
```

Comment:
Should we query the db here as well?

Reply:
I added the "ON CONFLICT IGNORE" to avoid having to do so. would like to avoid 2 likely round trips (most insertions should only be for new drvs, because we check if they are already present)